modules/apps/frontend-taglib/frontend-taglib/src/main/java/com/liferay/frontend/taglib/servlet/taglib/LogoSelectorTag.java

```diff
@@ -168,15 +185,66 @@ private String _getLogoURL(HttpServletRequest httpServletRequest) {
 		return getCurrentLogoURL();
 	}
 
-	private String _getSelectLogoURL(
-		HttpServletRequest httpServletRequest, String randomNamespace) {
+	private String _getPortletNamespace(HttpServletRequest httpServletRequest) {
+		return PortalUtil.getLiferayPortletResponse(
+			_getPortletResponse(httpServletRequest)
+		).getNamespace();
+	}
+
+	private PortletResponse _getPortletResponse(
+		HttpServletRequest httpServletRequest) {
 
 		PortletResponse portletResponse =
 			(PortletResponse)httpServletRequest.getAttribute(
 				JavaConstants.JAVAX_PORTLET_RESPONSE);
 
+		if (portletResponse != null) {
+			return portletResponse;
+		}
+
+		try {
+			Portlet portlet = PortletLocalServiceUtil.getPortletById(
+				PortletKeys.IMAGE_UPLOADER);
+
+			InvokerPortlet invokerPortlet = PortletInstanceFactoryUtil.create(
+				portlet, httpServletRequest.getServletContext());
+
+			PortletConfig portletConfig = PortletConfigFactoryUtil.create(
+				portlet, httpServletRequest.getServletContext());
+
+			ThemeDisplay themeDisplay =
+				(ThemeDisplay)httpServletRequest.getAttribute(
+					WebKeys.THEME_DISPLAY);
+
+			LiferayRenderRequest liferayRenderRequest =
+				RenderRequestFactory.create(
+					httpServletRequest, portlet, invokerPortlet,
+					portletConfig.getPortletContext(), WindowState.NORMAL,
+					PortletMode.VIEW,
+					PortletPreferencesFactoryUtil.fromDefaultXML(
+						portlet.getDefaultPreferences()),
+					themeDisplay.getPlid());
+
+			portletResponse = RenderResponseFactory.create(
+				themeDisplay.getResponse(), liferayRenderRequest);
+		}
+		catch (PortletException portletException) {
+			if (_log.isDebugEnabled()) {
+				_log.debug("Unable to get Portlet response", portletException);
+			}
+
+			portletResponse = RenderResponseFactory.create();
+		}
```

Comment:
If I am understanding this correctly, the main point is not to use `httpServletRequest` with hardcoded portlet id, instead of  `portletResponse`. If that is correct, there may be a simpler solutions, see
1. [solution in commerce](https://github.com/liferay/liferay-portal/blob/master/modules/apps/commerce/commerce-order-content-web/src/main/java/com/liferay/commerce/order/content/web/internal/fragment/renderer/OrderActionsFragmentRenderer.java#L264-L268), looks pretty simple 
2. what Ivan did [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/frontend/data/set/provider/CETFDSActionProvider.java#L130).

Reply:
Are you proposing to create the resource URL directly? That might work, I have not tried that approach.

Definitely looks simpler, but at the same time, we would loose the possibility of getting the portlet namespace like this:
```
	private String _getPortletNamespace(HttpServletRequest httpServletRequest) {
		return PortalUtil.getLiferayPortletResponse(
			_getPortletResponse(httpServletRequest)
		).getNamespace();
	}
```

We can overcome this by just using the porltet key though

I'm not 100% sure it'll work as the `httpServletRequest` we 'll be passing to the `PortletURLFactoryUtil.create()` method does not contain any portlet.

As this would largely simplify code, we can try.

Reply:
Yeah, instead of building "fake" `portletResponse`, the idea would be to build URL in a different way, with `PortletURLFactoryUtil.create`. I'm also not sure it it'll work, but the fact that is has portlet ID as an agrument makes me think it might.

Alternative to `PortletURLFactoryUtil` is `RequestBackedPortletURLFactory`, that Ivan is using [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/frontend/data/set/provider/CETFDSActionProvider.java#L201-L207).

Reply:
Pls see extra commit